types/src/nmt.rs

```diff
@@ -4,38 +4,76 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
 use crate::{Error, Result};
 
-const NS_ID_SIZE: usize = 29;
-const NS_ID_V0_MAX_SIZE: usize = 10;
+pub const NS_VER_SIZE: usize = 1;
+pub const NS_ID_SIZE: usize = 28;
+pub const NS_SIZE: usize = NS_VER_SIZE + NS_ID_SIZE;
+pub const NS_ID_V0_SIZE: usize = 10;
 
-pub type NamespacedHash = nmt_rs::NamespacedHash<NS_ID_SIZE>;
-pub type Nmt = nmt_rs::NamespaceMerkleTree<
-    MemDb<NamespacedHash>,
-    nmt_rs::NamespacedSha2Hasher<NS_ID_SIZE>,
-    NS_ID_SIZE,
->;
+pub type NamespacedHash = nmt_rs::NamespacedHash<NS_SIZE>;
+pub type NamespacedSha2Hasher = nmt_rs::NamespacedSha2Hasher<NS_SIZE>;
+pub type NamespaceProof = nmt_rs::nmt_proof::NamespaceProof<NamespacedSha2Hasher, NS_SIZE>;
+pub type Nmt = nmt_rs::NamespaceMerkleTree<MemDb<NamespacedHash>, NamespacedSha2Hasher, NS_SIZE>;
 
 #[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
-pub struct Namespace(nmt_rs::NamespaceId<NS_ID_SIZE>);
+pub struct Namespace(nmt_rs::NamespaceId<NS_SIZE>);
 
 impl Namespace {
+    pub fn from_raw(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() < NS_SIZE {
+            return Err(Error::InvalidNamespaceSize);
+        }
+
+        Namespace::new(bytes[0], &bytes[1..])
+    }
+
     pub fn new(version: u8, id: &[u8]) -> Result<Self> {
         match version {
             0 => Self::new_v0(id),
+            255 => Self::new_v255(id),
             n => Err(Error::UnsupportedNamespaceVersion(n)),
         }
     }
 
     pub fn new_v0(id: &[u8]) -> Result<Self> {
-        if id.len() > NS_ID_V0_MAX_SIZE {
-            return Err(Error::InvalidNamespaceIdSize(id.len()));
+        let id_pos = match id.len() {
+            // Allow 28 bytes len
+            NS_ID_SIZE => NS_ID_SIZE - NS_ID_V0_SIZE,
+            // Allow 18 bytes len or less
+            n if n <= NS_ID_V0_SIZE => 0,
+            // Anything else is an error
+            _ => return Err(Error::InvalidNamespaceSize),
+        };
+
+        let prefix = &id[..id_pos];
+        let id = &id[id_pos..];
+
+        // Validate that prefix is all zeros
+        if prefix.iter().any(|&x| x != 0) {
+            return Err(Error::InvalidNamespaceV0);
         }
 
-        let mut bytes = [0u8; NS_ID_SIZE];
-        bytes[NS_ID_SIZE - id.len()..].copy_from_slice(id);
+        let mut bytes = [0u8; NS_SIZE];
+        bytes[NS_SIZE - id.len()..].copy_from_slice(id);
 
         Ok(Namespace(nmt_rs::NamespaceId(bytes)))
     }
 
+    fn new_v255(id: &[u8]) -> Result<Self> {
```

Comment:
I'm not sure if I'm for `new_v255` name. I'd name it `try_max` because it just does validation and returns the `max()` namespace. The name `new_v255` may indicate there is some actual 255 version, where it's just a special constant like `-1` in C, where all bytes have to be 255, not just version.

Reply:
`try_max` sounds good